server/models/services/ModelService.js

```diff
@@ -132,28 +150,149 @@ const get = async (
  * @param {object} queryParams - query to be used to find the Sequelize Model to be updated
  * @param {object} updateParams - values to be updated when the Sequelize Model is found
+ * @param {object} options - Generic options for Sequelize
+ * @param {*} options.transaction - Sequelize transaction
  * @returns {Promise<*>} - result of the sequelize.update function
  */
-const update = async (model, queryParams, updateParams) => {
+const update = async (model, queryParams, updateParams, options = {}) => {
     if (!model) throw new Error('Model not defined');
 
     return await model.update(
         { ...updateParams },
-        { where: { ...queryParams } }
+        { where: { ...queryParams }, ...options }
     );
 };
 
+/**
+ * Gets or creates multiple records for multiple models as part of a
+ * single transaction.
+ *
+ * Accepts an array of objects specifying the get and create functions for the
+ * model, the values to get or create, and the attributes to use when returning
+ * the final result.
+ *
+ * If you need the result of previous operations, e.g. the ids, you can pass a
+ * function instead of an object.
+ *
+ * If you need to update fields but do not want to use them in the get query,
+ * you can specify an update function and updateValues.
+ *
+ * Returns an array of results, where each item is an array containing the model
+ * instance and a boolean for whether it was newly created.
+ *
+ * @example
+ * const results = await ModelService.nestedGetOrCreate([
+ *     {
+ *         get: getAtVersions,
+ *         create: createAtVersion,
+ *         values: { atId, atVersion },
+ *         returnAttributes: [null, []]
+ *     },
+ *     {
+ *         get: getBrowserVersions,
+ *         create: createBrowserVersion,
+ *         values: { browserId, browserVersion },
+ *         returnAttributes: [null, []]
+ *     },
+ *     {
+ *         get: getTestPlanTargets,
+ *         create: createTestPlanTarget,
+ *         values: { atId, browserId, atVersion, browserVersion },
+ *         returnAttributes: [null]
+ *     },
+ *     accumulatedResults => {
+ *         const [testPlanTarget] = accumulatedResults[2];
+ *         return {
+ *             get: getTestPlanReports,
+ *             create: createTestPlanReport,
+ *             update: updateTestPlanReport,
+ *             values: {
+ *                 testPlanTargetId: testPlanTarget.id,
+ *                 testPlanVersionId: testPlanVersionId
+ *             },
+ *             updateValues: { status },
+ *             returnAttributes: [null, [], [], [], []]
+ *         };
+ *     }
+ * ]);
+ *
+ * const [
+ *     [atVersion, isNewAtVersion],
+ *     [browserVersion, isNewBrowserVersion],
+ *     [testPlanTarget, isNewTestPlanTarget],
+ *     [testPlanReport, isNewTestPlanReport],
+ * ] = results;
+ *
+ * @param {[object|function]} getOptionsArray
+ * @returns {Promise<[[*,Boolean]]>}
+ */
+const nestedGetOrCreate = async getOptionsArray => {
```

Comment:
This is a pretty neat way to control what was needed

Reply:
Thanks 😁